File: streampark-spark/streampark-spark-connector/streampark-spark-connector-kafka/src/main/scala/org/apache/streampark/spark/connector/kafka/offset/MySQLOffset.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.streampark.spark.connector.kafka.offset

import org.apache.kafka.common.TopicPartition
import org.apache.spark.SparkConf
import scalikejdbc.{ConnectionPool, DB, _}

/**
 * MySQL Offset Manager
 *
 * The table model for storing offsets is as follow, (topic+groupId+partition as union unique
 * primary key)
 *
 * | ---------------------------------------------------------------------------------- |             |           |        |
 * |:-----------------------------------------------------------------------------------|:------------|:----------|:-------|
 * | topic                                                                              | groupId     | partition | offset |
 * | ---------------------------------------------------------------------------------- |             |           |        |
 * | topic_001                                                                          | groupId_001 | 0         | 197    |
 * | ---------------------------------------------------------------------------------- |             |           |        |
 * | topic_001                                                                          | groupId_001 | 1         | 200    |
 * | ---------------------------------------------------------------------------------- |             |           |        |
 */
private[kafka] class MySQLOffset(val sparkConf: SparkConf) extends Offset {

  private lazy val jdbcURL = storeParams("mysql.jdbc.url")
  private lazy val table = storeParams("mysql.table")
  private lazy val user = storeParams("mysql.user")
  private lazy val password = storeParams("mysql.password")

  /**
   * get stored offset
   *
   * @param groupId
   * @param topics
   * @return
   */
  override def get(groupId: String, topics: Set[String]): Map[TopicPartition, Long] = {
    require(topics.nonEmpty)
    ConnectionPool.singleton(jdbcURL, user, password)
    DB.getTable(table) match {
      case None =>
        DB.autoCommit {
          implicit session =>
            val sql =
              s"""
                 |create table if not exists $table (
                 |`topic` varchar(255),
                 |`groupId` varchar(255),
                 |`partition` varchar(255),
                 |`offset` int(10),
                 |UNIQUE INDEX `INX`(`topic`, `groupId`,`partition`)
                 |)
          """.stripMargin
            SQL(sql).execute.apply()
        }
        logWarn(s"storeType:MySQL,table: $table is not exist,auto created...")
        Map.empty[TopicPartition, Long]
      case Some(_) =>
        DB.readOnly {
          implicit session =>
            val where = topics.size match {
              case 1 => s""" `topic` = "${topics.head}"  """
              case _ =>
                s""" `topic` in (${topics.mkString("\"", "\",\"", "\"")}) """
            }
            val sql =
              s"select `topic`,`partition`,`offset` from $table where `groupId`=? and $where"
            SQL(sql)
              .bind(groupId)
              .map {
                result =>
                  new TopicPartition(result.string(1), result.int(2)) -> result
                    .long(3)
              }
              .list
              .apply()
              .toMap
        }
    }
  }

  /**
   * update offset
   *
   * @param groupId
   * @param offsetInfos
   */
  override def update(groupId: String, offsetInfos: Map[TopicPartition, Long]): Unit = {
    DB.localTx {
      implicit session =>
        offsetInfos.foreach {
          case (tp, offset) =>
            val sql =
              s"insert into $table(`topic`,`groupId`,`partition`,`offset`) values(?,?,?,?) on duplicate key update `offset`= values(`offset`) "
            val updated = SQL(sql)
              .bind(tp.topic(), groupId, tp.partition(), offset)
              .update()
            if (updated == 0) {
              throw new Exception(s"Commit kafka topic :${tp.topic()} failed!")
            }
            logInfo(s"storeType:MySQL,updateOffsets [ $groupId,${offsetInfos.mkString(",")} ]")
        }
    }
  }

  /**
   * delete offset
   *
   * @param groupId
   * @param topics
   */
  override def delete(groupId: String, topics: Set[String]): Unit = {
    DB.autoCommit {
      implicit session =>
        topics.foreach(topic => {
          val sql = "delete from $table where topic=? and groupId=?"
          SQL(sql).bind(topic, groupId).update()
        })
    }
    logInfo(s"storeType:MySQL,deleteOffsets [ $groupId,${topics.mkString(",")} ]")
  }
}

File: streampark-spark/streampark-spark-connector/streampark-spark-connector-kafka/src/main/java/org/apache/streampark/spark/connector/kafka/offset/MySQLOffset.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.spark.connector.kafka.offset;

import org.apache.streampark.common.conf.ConfigKeys;
import org.apache.streampark.common.util.JdbcUtils;

import org.apache.kafka.common.TopicPartition;
import org.apache.spark.SparkConf;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Set;
import java.util.stream.Collectors;

/** MySQL offset manager. */
class MySQLOffset extends Offset {

    private final String jdbcURL;
    private final String table;
    private final String user;
    private final String password;
    private final Properties jdbcConfig;

    MySQLOffset(SparkConf sparkConf) {
        super(sparkConf);
        Map<String, String> params = getStoreParams();
        jdbcURL = params.get("mysql.jdbc.url");
        table = params.get("mysql.table");
        user = params.get("mysql.user");
        password = params.get("mysql.password");
        jdbcConfig = new Properties();
        jdbcConfig.setProperty(ConfigKeys.KEY_ALIAS(), "spark-kafka-offset");
        jdbcConfig.setProperty("jdbcUrl", jdbcURL);
        jdbcConfig.setProperty("username", user);
        jdbcConfig.setProperty("password", password);
    }

    @Override
    public Map<TopicPartition, Long> get(String groupId, Set<String> topics) {
        if (topics.isEmpty()) {
            throw new IllegalArgumentException("topics must not be empty");
        }
        String where;
        if (topics.size() == 1) {
            where = " `topic` = \"" + topics.iterator().next() + "\" ";
        } else {
            where =
                    " `topic` in ("
                            + topics.stream()
                                    .map(t -> "\"" + t + "\"")
                                    .collect(Collectors.joining(","))
                            + ") ";
        }
        String sql =
                "select `topic`,`partition`,`offset` from "
                        + table
                        + " where `groupId`='"
                        + groupId
                        + "' and "
                        + where;
        List<Map<String, Object>> rows = JdbcUtils.select(sql, jdbcConfig);
        if (rows.isEmpty()) {
            return Collections.emptyMap();
        }
        Map<TopicPartition, Long> result = new HashMap<>();
        for (Map<String, Object> row : rows) {
            result.put(
                    new TopicPartition(String.valueOf(row.get("topic")), Integer.parseInt(String.valueOf(row.get("partition")))),
                    Long.parseLong(String.valueOf(row.get("offset"))));
        }
        return result;
    }

    @Override
    public void update(String groupId, Map<TopicPartition, Long> offsetInfos) {
        for (Map.Entry<TopicPartition, Long> entry : offsetInfos.entrySet()) {
            TopicPartition tp = entry.getKey();
            String sql =
                    "insert into "
                            + table
                            + "(`topic`,`groupId`,`partition`,`offset`) values('"
                            + tp.topic()
                            + "','"
                            + groupId
                            + "','"
                            + tp.partition()
                            + "','"
                            + entry.getValue()
                            + "') on duplicate key update `offset`= values(`offset`) ";
            int updated = JdbcUtils.update(sql, jdbcConfig);
            if (updated == 0) {
                throw new RuntimeException("Commit kafka topic :" + tp.topic() + " failed!");
            }
        }
        log.info("storeType:MySQL,updateOffsets [ {},{} ]", groupId, offsetInfos);
    }

    @Override
    public void delete(String groupId, Set<String> topics) {
        for (String topic : topics) {
            JdbcUtils.update(
                    "delete from " + table + " where topic='" + topic + "' and groupId='" + groupId + "'",
                    jdbcConfig);
        }
        log.info("storeType:MySQL,deleteOffsets [ {},{} ]", groupId, topics);
    }
}
